tanjun/dependencies/reloader.py

```diff
@@ -0,0 +1,388 @@
+        for path, value in scan_result.py_paths.items():
+            if tracked_value := self._waiting_for_py.get(path):
+                if value.last_modified_at == tracked_value:
+                    del self._waiting_for_py[path]
+                    await self._load_module(client, path)
+                    self._py_paths[path] = _PyPathInfo(value.sys_path, last_modified_at=value.last_modified_at)
+
+                else:
+                    assert value.last_modified_at is not None
+                    self._waiting_for_py[path] = value.last_modified_at
+
+            elif not (path_info := self._py_paths.get(path)) or path_info.last_modified_at != value.last_modified_at:
+                assert value.last_modified_at is not None
+                self._waiting_for_py[path] = value.last_modified_at
+
+        for path, value in scan_result.sys_paths.items():
+            if tracked_value := self._waiting_for_sys.get(path):
+                if value == tracked_value:
+                    del self._waiting_for_sys[path]
+                    await self._load_module(client, path)
+                    self._sys_paths[path] = value
+
+                else:
+                    self._waiting_for_sys[path] = tracked_value
+
+            elif self._sys_paths.get(path) != value:
+                self._waiting_for_sys[path] = value
+
+        if self._unload_on_delete:
+            for path in scan_result.removed_py_paths:
+                self._unload_module(client, path)
+
+                try:
+                    del self._py_paths[path]
+                except KeyError:
+                    pass
+
+            for path in scan_result.removed_sys_paths:
+                self._unload_module(client, path)
+
+                try:
+                    del self._sys_paths[path]
+                except KeyError:
+                    pass
+
+    @utilities.print_task_exc("Reloader crashed")
+    async def _loop(self, client: tanjun.Client, /) -> None:
+        while True:
+            await asyncio.sleep(self._interval)
+            await self.scan(client)
+
+    def stop(self) -> None:
+        if not self._task:
+            raise RuntimeError("Reloader not running")
+
+        self._task.cancel()
+        self._task = None
+
+    def start(self, client: alluka.Injected[tanjun.Client]) -> None:
+        if self._task:
+            raise RuntimeError("Reloader already started")
+
+        self._task = asyncio.create_task(self._loop(client))
+
+
+def _to_namespace(namespace: str, path: pathlib.Path) -> str:
+    return namespace + "." + path.name.removesuffix(".py")
+
+
+def _add_directory(
+    directory: typing.Union[str, pathlib.Path], namespace: typing.Optional[str]
+) -> tuple[pathlib.Path, list[pathlib.Path]]:
+    directory = pathlib.Path(directory)
+    if not directory.exists():
+        raise FileNotFoundError(f"{directory} does not exist")
+
+    return directory.resolve(), list(map(pathlib.Path.resolve, directory.glob("*.py")))
```

Comment:
Decide on the behaviour for symlinks:

1. When namespacing should it use the name of the symlink or the name of the file it links to or are these the same thing?
2. Should it unload the module when the symlink is deleted or only when the actual file is deleted?

Reply:
this behaviour is considered UB